### ISSUE TITLE
Pickling lambda function that accepts a global as argument is broken on windows

### ISSUE THREAD

**Issue by TheButlah** [cloudpipe/cloudpickle#230]
I tried to pickle a lambda function that calls another function with a global variable as an argument. It then unpickles the lambda in a new process and tries to call it. When it does so, it doesn't see the value of the global variable when it was pickled, but instead the original value of the variable, `None`. The issues goes away when I no longer modify the global variable using the global keyword in the helper function. The following is pseudocode of what fails:
```python3
args = None

def parse_args():
    # set up argument parser here (code omitted)

    global args
    args = parser.parse_args()

def main():
    parse_args()
    fns = [lambda: something(args)]*10
    wrappers = [CloudpickleWrapper(f) for f in fns]
    # For each wrapped function, start a process that calls that function (code omitted)
    # In these processes, they unexpectedly see `args` as being `None` instead of the results of the argument parser

if __name__=="__main__":
    main()
```

I am running Windows 10, Anaconda Python 3.6. **This issue did not occur on Ubuntu when I tested it!**. 

The following tests fail under windows when I run `tox`. **No tests failed under Ubuntu.** These are the tests:

```
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_closed_file PASSED
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_empty_file PASSED
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_pickling_special_file_handles PASSED
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_plus_mode PASSED
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_r_mode PASSED
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_seek PASSED
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_temp_file SKIPPED
tests/cloudpickle_file_test.py::CloudPickleFileTests::test_w_mode PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_Ellipsis PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_EllipsisType PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_NoneType PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_NotImplemented PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_NotImplementedType PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_abc PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_attrgetter PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_buffer PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_builtin_function_without_module PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_builtin_type__new__ PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_cell_manipulation PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_classmethod  PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_closure_interacting_with_a_global_variable PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_closure_none_is_preserved PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_correct_globals_import PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_cycle_in_classdict_globals PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_dynamic_module FAILED
tests/cloudpickle_test.py::CloudPickleTest::test_dynamic_modules_globals FAILED
tests/cloudpickle_test.py::CloudPickleTest::test_dynamic_pytest_module PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_dynamically_generated_class_that_uses_super PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_empty_cell_preserved PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_extended_arg PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_faulty_module PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_func_globals PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_function_from_dynamic_module_with_globals_modifications FAILED
tests/cloudpickle_test.py::CloudPickleTest::test_function_module_name PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_function_pickle_compat_0_4_0 SKIPPED
tests/cloudpickle_test.py::CloudPickleTest::test_function_pickle_compat_0_4_1 SKIPPED
tests/cloudpickle_test.py::CloudPickleTest::test_function_qualname PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_generator PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_import FAILED
tests/cloudpickle_test.py::CloudPickleTest::test_instancemethods_without_self PASSED
tests/cloudpickle_test.py::CloudPickleTest::test_interactively_defined_function FAILED
```

**Comment by pierreglaser** [comment]
@TheButlah the test suite should now work successfully on windows. Can you complete your  example (i.e include the code you use to start a process) so that we can investigate your issue?

**Comment by ogrisel** [comment]
FYI I have also fixed the tests under Python 2.7 on windows and added Windows jobs on our CI.

**Comment by TheButlah** [comment]
@pierreglaser my code is rather long and complex, im spawning multiple OpenAI Gym environments using the multiprocessing module, so my original code probably is not suitable for a unit test. What should I do instead?

**Comment by pierreglaser** [comment]
What function do you use to launch the new processes?

**Comment by TheButlah** [comment]
The code using multiprocessing is here: https://github.com/CornellDataScience/RL-Halite/blob/0b17e8ef29635f41978b8eaa9d6cfb0d1b516b29/utils/batched_env.py#L104-L136

The functions passed in are from here:   
`envs = BatchedEnv([lambda: gym.make(args.env_name)]*args.batch_size)`

**Comment by TheButlah** [comment]
The multithreading may be a red herring, it's probably just not picking correctly regardless of threading

**Comment by pierreglaser** [comment]
Here is what is happening:
We changed a couple month ago the handling of global variables in child processes. 

In short, if a global variable exists in your child process, and a new value for this variable is shipped within the pickle string of a function (which `cloudpickle` will do for `fns` and `args` for example), the already existing variable will have the priority, and override the one shipped by the function. 

In your case, the global variable `args` already exists in the child process, because on windows, `multiprocessing.Process` executes your file until `if __name__ === "__main__"`  and therefores sets `args` to `None`. When your function is depickled, cloudpickle will choose to stick to the current value of `args` instead of changing it with the value the pickle string of `f` contains.

This choice makes sense from a multiprocessing-library-developer point of view, but it seems to understandably surprise users. We are currently implementing a switch so that user can controls this behavior. See #216. Meanwhile, the solution you mention sounds like a good fix.

PS: the discussion does not hold for linux, because the default start method for new processes is ``fork``, which does not require python code execution, but rather cloning the current state of the parent process.

**Comment by ogrisel** [comment]
This should be fixed by #240.

**Comment by ogrisel** [comment]
The fix is now available in cloudpickle 0.8.0 on PyPI.
